ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -42,54 +42,131 @@
+  rb_event_flag_t current_state; // Used to coalesce similar events
+  bool sleeping; // Used to track when a thread is sleeping
+} thread_meta_t;
+
 static VALUE tracing_start(VALUE _self, VALUE output_path);
 static VALUE tracing_stop(VALUE _self);
+static VALUE tracing_clean(VALUE _self, VALUE thread);
+static VALUE use_object_id(VALUE _self);
 static double timestamp_microseconds(void);
-static void set_native_thread_id(void);
-static void render_event(const char *event_name);
+static void render_event(const char *event_name, thread_meta_t *thread);
 static void on_thread_event(rb_event_flag_t event, const rb_internal_thread_event_data_t *_unused1, void *_unused2);
 static void on_gc_event(VALUE tpval, void *_unused1);
 static VALUE mark_sleeping(VALUE _self);
-
-// Thread-local state
-static _Thread_local bool current_thread_seen = false;
-static _Thread_local unsigned int current_thread_serial = 0;
-static _Thread_local uint64_t thread_id = 0;
-static _Thread_local rb_event_flag_t previous_state = 0; // Used to coalesce similar events
-static _Thread_local bool sleeping = false; // Used to track when a thread is sleeping
+static const char* get_event_name(rb_event_flag_t event_id);
 
 // Global mutable state
-static rb_atomic_t thread_serial = 0;
 static FILE *output_file = NULL;
 static rb_internal_thread_event_hook_t *current_hook = NULL;
 static double started_tracing_at_microseconds = 0;
 static int64_t process_id = 0;
 static VALUE gc_tracepoint = Qnil;
 
+#if RUBY_3_3_PLUS
+static int gvl_tracing_key = 0;
+#else
+static _Thread_local thread_meta_t thread_local_meta = {};
+static rb_atomic_t thread_serial = 0;
+static void set_native_thread_id(void);
+#endif
+
 void Init_gvl_tracing_native_extension(void) {
   rb_global_variable(&gc_tracepoint);
 
+#if RUBY_3_3_PLUS
+  gvl_tracing_key = rb_internal_thread_specific_key_create();
+#endif
   VALUE gvl_tracing_module = rb_define_module("GvlTracing");
 
   rb_define_singleton_method(gvl_tracing_module, "_start", tracing_start, 1);
   rb_define_singleton_method(gvl_tracing_module, "_stop", tracing_stop, 0);
+  rb_define_singleton_method(gvl_tracing_module, "_clean", tracing_clean, 1);
+  rb_define_singleton_method(gvl_tracing_module, "_use_object_id?", use_object_id, 0);
   rb_define_singleton_method(gvl_tracing_module, "mark_sleeping", mark_sleeping, 0);
 }
 
-static inline void initialize_thread_id(void) {
-  current_thread_seen = true;
-  current_thread_serial = RUBY_ATOMIC_FETCH_ADD(thread_serial, 1);
+#if RUBY_3_3_PLUS
+
+static thread_meta_t* current_thread_meta(VALUE thread) {
+  void *data = rb_internal_thread_specific_get(thread, gvl_tracing_key);
+  thread_meta_t* thread_meta = NULL;
+
+  if (data != NULL) {
+    thread_meta = (thread_meta_t *)data;
+  } else {
+    thread_meta = (thread_meta_t*)malloc(sizeof(thread_meta_t));
+    memset(thread_meta, 0, sizeof(thread_meta_t));
+    thread_meta->thread = thread;
+    thread_meta->thread_object_id = RB_NUM2INT(rb_obj_id(thread));
```

Comment:
Could [`Thread#native_thread_id`](https://ruby-doc.org/3.2.2/Thread.html#method-i-native_thread_id) work here?

Although there's a stipulation that might make it unusable here:

> [!NOTE]  
> If the thread is not associated yet or already deassociated with a native thread, it returns nil. If the Ruby implementation uses M:N thread model, the ID may change depending on the timing.

Reply:
Thanks for the input @amomchilov !

There are a couple obstacles, unfortunately. 

First, as far as I can tell there is no way to safely access the `native_thread_id` method from C without the GVL. The thread api exposed to C extension code is pretty limited, that is safe to run without the GVL. With the way this code works it is always updating the file directly during the hook events, which have no GVL guarantees. Part of me is hoping you'll tell me i'm wrong here and there is a way to invoke thread methods safely in this context :D 

Second is like you mentioned - ideally we'd solve M:N support here as well. The M:N model is actually even more strict than that note, it always returns nil in the current implementation.

Reply:
Makes sense! I’m just squinting and doing loose pattern matching here. I have no expertise in this domain :)

there rent all that many threads. Touching their object IDs would be totally fine.